src/charm.py

```diff
@@ -236,16 +240,82 @@ def _command(self) -> str:
+            args.append(f"--storage.tsdb.retention.time={retention_time}")
+
+        # Set time series retention size
+        try:
+            ratio = self._percent_string_to_ratio(config.get("maximum_retention_size", ""))
+
+        except ValueError as e:
+            logger.warning(e)
+            self.unit.status = BlockedStatus(f"Invalid retention size: {e}")
+
+        else:
+            # `storage.tsdb.retention.size` uses the legacy binary format, so "GB" and not "GiB"
+            # https://github.com/prometheus/prometheus/issues/10768
+            # For simplicity, always communicate to prometheus in GiB
+            try:
+                capacity = self._convert_k8s_capacity_to_legacy_binary_gigabytes(
+                    self._get_pvc_capacity(), ratio
+                )
+            except ValueError as e:
+                self.unit.status = BlockedStatus(f"Error calculating retention size: {e}")
+            except LightkubeApiError as e:
+                self.unit.status = BlockedStatus(
+                    f"Error calculating retention size "
+                    f"(try running `juju trust` on this application): {e}"
+                )
+            else:
+                logger.debug("Retention size limit set to %s (%s%%)", capacity, ratio * 100)
+                args.append(f"--storage.tsdb.retention.size={capacity}")
 
         command = ["/bin/prometheus"] + args
 
         return " ".join(command)
 
-    def _is_valid_timespec(self, timeval) -> bool:
+    def _convert_k8s_capacity_to_legacy_binary_gigabytes(self, capacity: str, multiplier) -> str:
+        # Reduce possible ambiguity in unit name by adding a trailing 'B'.
+        # (bitmath doesn't accept e.g. "Gi" because it also supports bits.)
+        if not capacity.endswith("B"):
+            capacity += "B"
+
+        # For simplicity, always communicate to prometheus in GiB
+        storage_value = multiplier * bitmath.parse_string(capacity).to_GiB().value
+        return f"{storage_value}GB"
+
+    def _get_pvc_capacity(self) -> str:
+        # Get PVC capacity from kubernetes
+        client = Client()
+        pod_name = self.unit.name.replace("/", "-", -1)
+
+        # Take the first volume whose name starts with "<app-name>-database-".
+        # The volumes array looks as follows for app "am" and storage "data":
+        # 'volumes': [{'name': 'am-data-d7f6a623',
+        #              'persistentVolumeClaim': {'claimName': 'am-data-d7f6a623-am-0'}}, ...]
+        pvc_name = ""
+        for volume in cast(
+            Pod, client.get(Pod, name=pod_name, namespace=self.model.name)
+        ).spec.volumes:
+            # TODO: How to handle this when the store is not a singleton (i.e. when metadata
+            #  specifies "multiple: range: 1-10")?
+            if volume.persistentVolumeClaim.claimName.startswith(f"{self.app.name}-database-"):
+                pvc_name = volume.persistentVolumeClaim.claimName
+                break
```

Comment:
@rbarry82 what would be the best way here to obtain the PVC?

Is multiple stores possible? I added the following to prometheus-k8s metadata:
```diff
 storage:
   database:
     type: filesystem
+    multiple:
+      range: 3-5
```
And when I deploy I see

`ERROR juju.worker.caasapplicationprovisioner.runner exited "prometheus-k8s": ensuring application "prometheus-k8s": duplicated storage name "database" for "prometheus-k8s" not valid`